DataScience/CookedLogSequence.py

```diff
@@ -0,0 +1,80 @@
+from datetime import datetime
+from loggers import Logger
+import re, os
+
+class CookedLogSequence:
+    timestamp_format = '%Y-%m-%dT%H:%M:%S.%f'
+    timestamp_pattern = r'[^\\]"Timestamp"\s*:\s*"(.*?)0Z"'
+    def __init__(self, files):
+        self.files = files
+    
+    def get_time_from_log_line(self, log_line):
+        return datetime.strptime(re.search(self.timestamp_pattern, log_line).group(1), self.timestamp_format)
+
+    def get_file_start_time(self, fp):
+        with open(fp, 'r') as f:
+            first_line = f.readline()
+            return self.get_time_from_log_line(first_line)
+    
+    def get_file_end_time(self, fp):
+
+        with open(fp, 'r') as f:
+            last_line = f.readlines()[-1]
+            return self.get_time_from_log_line(last_line)
+    
+    def get_first_ts(self):
+        if len(self.files) > 0:
+            return self.get_file_start_time(self.files[0])
+        return datetime.min
+    
+    def get_last_ts(self):
+        if len(self.files) > 0:
+            return self.get_file_end_time(self.files[-1])
+        return datetime.min
+    
+    def find_first_file_to_merge(self, files, merged_end_time):
+        if self.get_file_end_time(files[-1]) < merged_end_time:
+            return -1
+        candidate_index = len(files) - 1
+        for i, fp in enumerate(files):
+            if self.get_file_start_time(fp) > merged_end_time:
+                candidate_index = i
+                break
+        if candidate_index > 0 and self.get_file_end_time(files[candidate_index - 1]) > merged_end_time:
```

Comment:
putting this conditional outside the for loop allows us to read the last line of a file for only a single file instead of reading the last line for all files passed in. This was a solution addressing @ataymano commend about the cost of reading the last line in a file.

Reply:
if logs are valid, then second condition is redundant actually.

Reply:
Maybe it would be more readable if we put trivial cases in the beginning:
1) make it function of other: CookedLogSequence instead of other: list
2) this function is always called with merged_end_time == self.get_last_ts, so this parameter is redundant and can be removed
3) then both corner cases (in the beginning) can be written in the beginning and the function will be:
def _find_first_file_to_merge(self, other):
  last_ts = self.get_last_ts()
  if other.get_last_ts() < last_ts:
     return None
  if other.get_first_ts() >= last_ts:
     return 0
   # non-trivial case